Mpi/better initialization

jcotela opened the pull request:
This is a proposal for a more fine-grained control of MPI initialization. Included changes:
- The Kernel now knows if this is an MPI run or not (decided by the presence of the `--using-mpi` command-line argument). This can be accessed from C++ and Python through `Kernel.IsDistributedRun()`.
- Importing the mpi module is now separate from MPI initialization/finalization. MPI is initialized when importing `KratosMultiphysics` for the first time **if and only if** the `--using-mpi` flag is given.
- The MPIEnvironment, which manages MPI initialization and finalization, has been reimplemented as a Meyers singleton class. MPI Finalization now happens on the C++ level, during the destruction of the singleton (so at the end of program execution).
- The ParallelEnvironment class, which was already a (naive) singleton, has been made more robust with regards to common singleton lifetime pitfalls (it is now also impossible to destruct from python).
- Adding a function in kratos_utilities to check for availability of mpi (analogue to functions to check for application availability).
- Changes in the legacy python-mpi interface and the to adapt to the new behavior.
- Minor bugfix in the logger python interface.

The main benefit of the proposed changes is that the mpi module is safe to import in serial runs (it will not try to change the default logger or initialize mpi), so this PR should close #4731 and similar problems.

~The main drawback is that it is a behavior change: the `--using-mpi` flag is now MANDATORY for mpi runs, since there is no other way to force a `MPI_Init`.~
**Update:** with the latest version, an attempt is made to detect a parallel run through (implementation-dependent) MPI environment variables. The `--using-mpi` flag is now optional and should only be necessary to force MPI initialization in unsupported MPI implementations.

jcotela committed:
Adding kernel-level knowledge of mpi/serial run.

jcotela committed:
Adding mpi availability check in kratos_utilities.

jcotela committed:
Removing unneeded import.

jcotela committed:
Merging master.

jcotela committed:
Reverting unnecessary modification.

jcotela committed:
ParallelEnvironment now a proper Meyers singleton.

jcotela committed:
MPIEnvironment reimplemented as singleton.

jcotela committed:
Moving GetMPICommunicator method to MPIDataCommunicator.

jcotela committed:
Making mpi init explicit.

jcotela committed:
Lazy initialization of the legacy mpi python interface.
It prevents accessing the parallel environment before initialization.

jcotela committed:
Fixing bug with kwargs access and return types.

jcotela committed:
syntax error: missing self.

jcotela committed:
More descriptive error message in case of failure.

jcotela committed:
Removing commented out code.

jcotela committed:
Removing one-line function that is used only once.

jcotela committed:
Updating logger calls (since it is now mpi-aware).

jcotela committed:
Remove unused Rank request.

pooyan-dadvand reviewed (commented):
Is there any special reason to have the Kernel::IsDistributedRun non static?

RiccardoRossi commented on kratos/sources/kernel.cpp:
```diff
@@ -21,15 +21,15 @@
-Kernel::Kernel() : mpKratosCoreApplication(Kratos::make_shared<KratosApplication>(
-                std::string("KratosMultiphysics"))) {
+Kernel::Kernel(bool IsDistributedRun) : mpKratosCoreApplication(Kratos::make_shared<KratosApplication>(
+                std::string("KratosMultiphysics"))), mIsDistributedRun(IsDistributedRun) {
```
a minor comment, however i would propose to have after the logo an output tellign that we are running indistributed mode (otherwise it is hard to tell by just seeing the run)

RiccardoRossi reviewed (commented):
At first sight this looks good to me, nevertheless i am not clear of what is the guideline, for example within a test, to verify if we are in mpi mode or not.

aside of this i would have a minimal output to confirm that we are in distributed mode

jcotela commented:
@RiccardoRossi as a minimal guideline:
from python, check if the current run is mpi:
```python
if KratosMultiphysics.IsDistributedRun():
  #mpi stuff
```
alternatively, check if the mpi module is available for importing:
```python
import kratos_utilities
if kratos_utilities.IsMPIAvailable():
  import mpi # note that this no longer implies mpi initialization, so functions in this module might not work (and should error) in serial runs
```

jcotela commented:
@pooyan-dadvand I could not make the function static because it uses member information (the bool storing whether this is mpi or not). It is true that it might make sense to set this globally, perhaps placing this data in the `ParallelEnvironment` (which is already a singleton). Otherwise this information is hard to get from the C++ layer (one would need a kernel instance). What do you think?

philbucher commented:
> @RiccardoRossi as a minimal guideline:
> from python, check if the current run is mpi:
> 
> ```python
> if KratosMultiphysics.IsDistributedRun():
>   #mpi stuff
> ```
> 
> alternatively, check if the mpi module is available for importing:
> 
> ```python
> import kratos_utilities
> if kratos_utilities.IsMPIAvailable():
>   import mpi # note that this no longer implies mpi initialization, so functions in this module might not work (and should error) in serial runs
> ```

to me this makes perfect sense and is very self explaining, +2 from my side for this

philbucher commented:
> aside of this i would have a minimal output to confirm that we are in distributed mode

I suggest to output this in the same way as the compiled mode (Release, FullDebug,...):
~~~

 |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 7.0.0-f84c924-FullDebug - DistrubutedRun # name not fixed, just general suggestion
~~~

jcotela committed:
Removing unused function.

jcotela committed:
Adding parallel mode to banner.

jcotela commented:
```
 |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 7.0.0-7446cd4bb5 -- MPI parallel run
```
Something like this?

I also added a serial/openmp notification, which I am not completely happy with (should the MPI parallel run specify OpenMP support too?).

jcotela committed:
Merge branch 'master' into mpi/better-initialization

pooyan-dadvand commented:
> @pooyan-dadvand I could not make the function static because it uses member information (the bool storing whether this is mpi or not). It is true that it might make sense to set this globally, perhaps placing this data in the `ParallelEnvironment` (which is already a singleton). Otherwise this information is hard to get from the C++ layer (one would need a kernel instance). What do you think?

But the bool `mIsDistributedRun `can also be a static member. (or static local variable of the method)

pooyan-dadvand commented:
I would add to the version if is compiled with MPI/OpenMP 

And in another line if is started with `--using-mpi` or not. And what is the number of threads in OpenMP

RiccardoRossi commented:
thx about the mini guide.

i would have something like

MPI         world_size    =128 
OpenMp max_theads=8

jcotela committed:
Merging in recent changes in master.

jcotela committed:
More detailed parallelism reporting.

jcotela committed:
Making Kernel::IsDistributedRun static.

jcotela commented:
I added a more detailed parallel info printout and made `Kernel::IsDistributedRun` static. Some examples from a test file importing KratosMultiphysics:

`python test.py`:
```
 |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 7.0.0-549efcdaa6-Release
Compiled with OpenMP and MPI support.
Maximum OpenMP threads: 16.
Running without MPI.
```
`mpirun -np 2 python test.py`
``` |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 7.0.0-549efcdaa6-Release
Compiled with OpenMP and MPI support.
Maximum OpenMP threads: 2.
Running without MPI.
```
`mpirun -np 2 python test.py --using-mpi`
``` |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 7.0.0-1455582c26-Release
Compiled with OpenMP and MPI support.
Maximum OpenMP threads: 2.
MPI world size:         2.
```
`OMP_NUM_THREADS=1 mpirun -np 2 python test.py --using-mpi`
```
 |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 7.0.0-549efcdaa6-Release
Compiled with OpenMP and MPI support.
Maximum OpenMP threads: 1.
MPI world size:         2.
```

RiccardoRossi commented:
love it.

on my side i would approve but i let to pooyan since you are in discussion with him

philbucher commented:
how does it look if omp-support is explicitly disabled?

pooyan-dadvand reviewed (dismissed):
Thanks!

jcotela commented:
@philbucher This is my mpi-only test compilation (serial run):
```
 |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 7.0.0-d345ad9d58-FullDebug
Compiled with MPI support.
Running without MPI.
```
If I also disable MPI, I get:
```
 |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 7.0.0-d345ad9d58-FullDebug
Serial compilation.
```

jcotela commented:
Please do not merge yet, I just discovered a problem with tests. In particular, C++ tests for MPI features error out unless the `--using-mpi` option is given. While you could say that this means that they are working as intended, they are also always run when calling `Tester.RunAllTests()` and I do not know if there is a way to cleanly exclude them.

So far, all tests I've done for MPI are in `KratosMPICoreFastSuite` so an option could be to automatically disable that suite in non parallel runs, but I do not know if hard-coding behavior for one specific test suite name is a good idea.

I am open to suggestions... @pooyan-dadvand I believe you designed the C++ testing infrastructure, what should we do here?

By the way, I have seen that codacy complains about using `constexpr bool` as arguments within ifs... It is easy to fix it by replacing the if with macro #ifdef (I'm sure I can also figure out something with templates :smiling_imp:), but I seriously doubt that the resulting code is clearer. I can do it if desired, just let me know.

jcotela committed:
virtual (const) inquiry methods
to allow reimplementation in derived classes.

jcotela committed:
Adding base class for distributed test cases.

jcotela committed:
Defining macros to register distributed tests.

jcotela committed:
Using distributed macro for mpi-only tests.

pooyan-dadvand commented:
The first option I see is to add macros for MPI which checks the `Kernel::IsDistributed` and add or not the MPI tests. 

Another way is to add a new Environment variable to the testing mechanism to check about run, os, etc. This would be more work but more extensible

jcotela committed:
Merge pull request #4918 from KratosMultiphysics/mpi/distributed-test-case

Mpi/distributed test case

jcotela committed:
Merge branch 'master' into mpi/better-initialization

jcotela committed:
Auto-detection of MPI for initialization.

jcotela committed:
Removing extra Test in test name.

jcotela commented:
After discussing with @pooyan-dadvand and @philbucher, the test issue has been solved by introducing a new base class for distributed tests, which automatically disables the test when running without MPI.
Please use the `KRATOS_DISTRIBUTED_TEST_CASE_IN_SUITE` macro to define tests that require MPI.

I have also added capabilities for the automatic detection of MPI runs that should work for OpenMPI and Intel MPI implementations. I have also added support for detection of mpich-based MPI, but the environment variable that is checked is very recent (was added to the master of mpich a couple of months ago) and I do not expect it to be very reliable at the moment. This was originally proposed by @maceligueta some months ago and we decided not to do it back then. However, I think that the extra prints regarding MPI settings added in this PR should be enough to remove uncertainties in case of failure of the automated detection mechanism.

For MPI implementations not supported by the automated detection, the option remains to force MPI initialization through the `--using-mpi` flag remains available.

pooyan-dadvand commented on kratos/includes/parallel_environment.h:
```diff
@@ -142,6 +147,9 @@ class KRATOS_API(KRATOS_CORE) ParallelEnvironment
 
     std::unordered_map<std::string, DataCommunicator::UniquePointer>::iterator mDefaultCommunicator;
 
+    static ParallelEnvironment* mpInstance;
```
Why you have not used a Meyer singleton here?

jcotela replied:
I use the pointer to guard against a race condition if the first access even happens to be within an openmp parallel region, as well as to keep track of singleton destruction:
https://github.com/KratosMultiphysics/Kratos/blob/abd79e5d7fba2aaf3cff4b19aacac9d7934e5c76/kratos/sources/parallel_environment.cpp#L89-L113

The object itself is still a static (local) variable. This is all from the "recipe" in Modern C++, and it helped me to catch a bug due to interaction between the `ParallelEnvironment` and `MPIEnvironment` object lifetimes.

pooyan-dadvand commented on kratos/mpi/mpi_environment.h:
```diff
@@ -18,38 +18,46 @@
+
+    static void Create();
+
+    static MPIEnvironment* mpInstance;
```
Again an static member instance

pooyan-dadvand commented on kratos/includes/kernel.h:
```diff
@@ -70,7 +70,7 @@ class KRATOS_API(KRATOS_CORE) Kernel {
 
     */
-    Kernel();
+    Kernel(bool IsDistributedRun = false);
```
Why we need to expose this to all? It shouldn't be a protected  constructor for MPI kernel to set it correctly?

jcotela replied:
There is no MPI-specific version of the Kernel currently. I can add it as desired, but I did not do it at the moment, since the only difference between the two would be setting a static member variable to a different value.

pooyan-dadvand reviewed (commented):
Very nice work! Just few minor comments.

pooyan-dadvand reviewed (dismissed):
👍

jcotela committed:
Using IsMPIAvailable instead of custom check in init.

philbucher reviewed (approved):
perfect :+1:
